dessia_common/datatools/modeling.py

```diff
@@ -0,0 +1,734 @@
+"""
+Tools and base classes for machine learning methods.
+
+"""
+from typing import List, Dict, Any
+
+import numpy as npy
+from sklearn import preprocessing, linear_model, ensemble, tree, svm, neural_network
+
+from dessia_common.core import DessiaObject
+
+
+# ======================================================================================================================
+#                                                     S C A L E R S
+# ======================================================================================================================
+class BaseScaler(DessiaObject):
+    _rebuild_attributes = []
+
+    def __init__(self, name: str = ''):
+        DessiaObject.__init__(self, name=name)
+
+    @classmethod
+    def _skl_class(cls):
+        raise NotImplementedError('Method _skl_class not implemented for BaseScaler. Please use children.')
+
+    def _call_skl_scaler(self):
+        return self._skl_class()()
+
+    def instantiate_skl(self):
+        scaler = self._call_skl_scaler()
+        for attr in self._rebuild_attributes:
+            setattr(scaler, attr, getattr(self, attr))
+        return scaler
+
+    @classmethod
+    def instantiate_dessia(cls, scaler, name: str = ''):
+        kwargs_dict = {'name': name}
+        for attr in cls._rebuild_attributes:
+            if hasattr(scaler, attr):
+                if isinstance(getattr(scaler, attr), npy.ndarray):
+                    kwargs_dict[attr] = getattr(scaler, attr).tolist()
+                    continue
+                kwargs_dict[attr] = getattr(scaler, attr)
+        return cls(**kwargs_dict)
+
+    @classmethod
+    def fit(cls, matrix: List[List[float]], name: str = ''):
+        scaler = cls._skl_class()()
+        scaler.fit(matrix)
+        return cls.instantiate_dessia(scaler, name)
+
+    def transform(self, matrix: List[List[float]]):
+        scaler = self.instantiate_skl()
+        return scaler.transform(matrix).tolist()
+
+    @classmethod
+    def fit_transform(cls, matrix: List[List[float]], name: str = ''):
+        scaler = cls.fit(matrix, name)
+        return scaler, scaler.transform(matrix)
+
+
+class StandardScaler(BaseScaler):
+    """
+    Data scaler that standardly scale data. The operation made by this scaler is `new_X = (X - mean(X))/std(X)`.
+
+    :param mean_:
+        --------
+        List of means
+    :type mean_: `List[float]`, `optional`, defaults to `None`
+
+    :param scale_:
+        --------
+        List of standard deviations
+    :type mean_: `List[float]`, `optional`, defaults to `None`
+
+    :param var_:
+        --------
+        List of variances
+    :type mean_: `List[float]`, `optional`, defaults to `None`
+
+    """
+    _rebuild_attributes = ['mean_', 'scale_', 'var_']
+    _standalone_in_db = True
+
+    def __init__(self, mean_: List[float] = None, scale_: List[float] = None, var_: List[float] = None, name: str = ''):
```

Comment:
Why do we escape variable names with trailing underscores here ? Are these reserved by python in some way ?

Reply:
These variable names are the sklearn object attributes, which cannot be changed in sklearn objects.
For the sake of simplicity, i named them the same in Dessia Objects as in sklearn objects.

The whole purpose of this script is to pre-serialize sklearn objects so that they can be created, fit (thanks to sklearn methods) and stored in db as DessiaO and reused at another time by reinstantiating sklearn object from dessia objects attributes.